### PR TITLE
Add a linter.

### PR DESCRIPTION
**Pull request by SkySkimmer**
Tabs are an annoyance in Ocaml files as ocp-indent does not know how to deal with them. This PR attempts to work towards their removal without having a multi thousand line patch of nothing but whitespace changes.

Currently it enforces that all added lines in ocaml files should not start with a tab (here added is according to git, which doesn't have a concept of modified line). This is seems as weak a check as possible while working towards getting rid of tabs as indentation in ocaml files.

The main limitations are that:
- if a commit has remaining tabs they can be annoying to find and remove
- when doing a search-and-replace type of change (for instance like the ast change) one has to get rid of tabs on every changed line

This work can be done automatically by the Emacs package [ws-butler](https://github.com/lewang/ws-butler) (with `ws-butler-convert-leading-tabs-or-spaces` enabled). However this does not help anyone not using emacs or otherwise unwilling to install the package. Is there anyone who this applies to?

A script to automatically fix commits should also be possible (ie run as `git rebase --exec dev/tools/fix-tabs.sh`).

Does this seem like a reasonable requirement for PRs to everyone?

PS: ideas for other checks or to improve the tab check are welcome ;)

### CODE REVIEW

**Comment by mattam82** [comment]
I'd like to see how that one works. I think I'm guilty of introducing some, but I already use ws-buttler for end of lines, so it's gonna be easy to avoid the tabs at the beginning too :)

**Comment by SkySkimmer** [comment]
>I'd like to see how that one works.

What do you mean?

**Comment by mattam82** [comment]
I mean we can experiment if it ends up being too impractical change the way
we deal with this issue.

**Comment by Zimmi48** [comment]
Can you also add a check for trailing spaces in this linter?
And would it also be feasible to test that there is a new line at EOF? (As discussed in #986.)

**Comment by SkySkimmer** [comment]
Files without eol at eof can be detected with something like
``` bash
BADFILES=$(mktemp)

# super tricky: $() removes ending newlines
# so the condition tests for nonempty files ending with non-newlines
# see https://stackoverflow.com/a/5922741
# quoting is insane
COND_CMD='if [ -n "$'"(tail -c 1 '{}')"'" ]; then echo '"'{}' >> $BADFILES; fi"

find . "(" -path ./.git -prune ")" -o -type f -exec 'bash' '-c' "$COND_CMD" ';'

if [ -n "$(cat "$BADFILES")" ];
then
    >&2 echo "Files without ending newlines!"
    >&2 cat "$BADFILES"
fi
```
(the COND_CMD is horrible but could be put in a helper script)

This outputs:
```
bash: warning: command substitution: ignored null byte in input
bash: warning: command substitution: ignored null byte in input
Files without ending newlines!
./dev/v8-syntax/check-grammar
./dev/build/windows/configure_profile.sh
./dev/build/windows/patches_coq/coq-8.4pl2.patch
./dev/build/windows/patches_coq/ln.c
./dev/build/windows/MakeCoq_regtest_noproxy.bat
./dev/nsis/FileAssociation.nsh
./test-suite/.csdp.cache
./test-suite/success/guard.v
./test-suite/coq-makefile/timing/per-file-after/A.v.timing.diff.desired
./test-suite/coq-makefile/timing/after/time-of-build-both.log.desired
./test-suite/coqdoc/links.html.out
./test-suite/coqdoc/bug5648.html.out
./test-suite/coqdoc/bug5700.html.out
./test-suite/bugs/opened/3010.v-disabled
./test-suite/bugs/opened/3478.v-disabled
./test-suite/bugs/opened/3186.v-disabled
./test-suite/bugs/closed/5245.v
./doc/common/styles/html/simple/style.css
./doc/refman/coqide.png
./doc/refman/index.html
./doc/refman/coqide-queries.png
./ide/MacOS/coqide.icns
./ide/coq.png
./ide/coq2.ico
```
So it would have to be filtered in some way.

**Comment by SkySkimmer** [comment]
I pushed a new version based on `git diff --check` and gitattributes which checks trailing spaces. It also prints where and what the whitespace errors are. With a dummy repo:
```
thefile:3: tab in indent.
+       fdlknkn
thefile:14: tab in indent.
+       lknl
```

**Comment by SkySkimmer** [comment]
Added the end of file newline check.

I think this is good to merge.
If there are no objections in the next 2~ days I'll remove the need discussion/feedback labels.

**Review by Zimmi48 — APPROVED** [review]
Great work! Thanks @SkySkimmer.

File: .travis.yml [hunk 1]
```diff
@@ -135,6 +135,17 @@ matrix:
...
+      - TEST_TARGET="lint"
+      install: []
+      before_script: []
+      addons:
+      apt:
+        sources: []
+        packages: []
+      script:
+        - dev/lint-repository.sh
+
```
[the remark this answers is not in the document]
Comment (SkySkimmer) [reply .travis.yml]: I think I just put it at the end of the linux tests.
Feel free to push a reorder to my branch.

[the remark this answers is not in the document]
Comment (Zimmi48) [reply .travis.yml]: Hum, I only put it a bit higher in the list, which isn't better than what was done before. I don't know how to put it at the beginning. Feel free to get rid of my commit.

**Comment by SkySkimmer** [comment]
>dev/build/windows/MakeCoq_regtest_noproxy.bat:28: trailing whitespace

Any windows experts know if I need CRLF at the end of the file or is LF enough? Or we could just disable the whitespace check for .bat files.

**Comment by Zimmi48** [comment]
> Or we could just disable the whitespace check for .bat files.

That wouldn't be a big deal to disable all these checks on `.bat` files IMHO.

**Comment by SkySkimmer** [comment]
OK then.

**Comment by SkySkimmer** [comment]
Actually we can make git treat CRLF correctly in bat files so I did that.

**Comment by SkySkimmer** [comment]
Travis is being stupid for whatever reason.

Anyway if there are no objections I'll remove the need:discussion tags Tuesday.

**Comment by SkySkimmer** [comment]
As far as I'm concerned this is ready to merge.

**Pull request merged**